### PR TITLE
[Data] Use runtime object memory for scheduling

### LINKED ISSUE #41190: [data] Add object_store_memory to `incremental_resource_usage()`
Backpressure master issue: #40754.

`incremental_resource_usage()` currently doesn’t include object store memory (i.e., size of task outputs). This means that we only check if the current object store memory already exceeds the global limit, but don’t consider if submitting a new task will make the usage exceed the limit.

Consider a case where CPUs/GPUs are far from full, but object store memory is close to the limit. In this case, the executor will still quickly submit many tasks, and then lead to OOM.

We can update the object_store_memory resource in `incremental_resource_usage()` based on the metrics collected from finished tasks. We already have this information in `OpRuntimeMetrics`.

To update the object_store_memory resource, we can use 1) either the total output block size of a task ; 2) or streaming generator backpressure buffer size (if StreamingOutputs backpressure is enabled).

### PR DESCRIPTION
**Pull request by bveeramani**
## Why are these changes needed?

When selecting an operator to run, the scheduler doesn't consider how much object store memory an operator consumes. If an operator produces large blocks, the scheduler might select the operator too frequently, and your cluster can out of memory.

To prevent this from happening, this PR updates the scheduler so that it considers the incremental object store memory usage when selecting an operator.

## Related issue number

Fixes https://github.com/ray-project/ray/issues/41190

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/execution/streaming_executor_state.py [hunk 1]
```diff
@@ -608,4 +604,11 @@ def _execution_allowed(
         object_store_memory=downstream_usage.object_store_memory
     ).satisfies_limit(downstream_limit)
 
+    if (
+        global_ok_sans_memory
+        and inc.object_store_memory is not None
+        and inc.object_store_memory <= 0
+    ):
+        return True
```
Comment (bveeramani) [inline python/ray/data/_internal/execution/streaming_executor_state.py:612]: We could split this off into a separate PR.

File: python/ray/data/context.py [hunk 2]
```diff
@@ -102,6 +102,11 @@
     int(os.environ.get("RAY_DATA_USE_STREAMING_EXECUTOR", "1"))
 )
 
+# Whether to use the runtime object store memory metrics for scheduling.
+DEFAULT_USE_RUNTIME_METRICS_SCHEDULING = bool(
+    int(os.environ.get("DEFAULT_USE_RUNTIME_METRICS_SCHEDULING", "1"))
```
Comment (bveeramani) [inline python/ray/data/context.py:107]: Will set this to `0` before merging. Currently enabled to ensure tests pass.

File: python/ray/data/_internal/execution/streaming_executor_state.py [hunk 3]
```diff
@@ -645,4 +648,14 @@ def _execution_allowed(
...
+    # even if we're over the global limit.
+    if (
+        DataContext.get_current().use_runtime_metrics_scheduling
+        and global_ok_sans_memory
+        and op.metrics.average_bytes_change_per_task is not None
+        and op.metrics.average_bytes_change_per_task <= 0
+    ):
+        return True
+
     return global_ok_sans_memory and downstream_memory_ok
```
Comment (raulchen) [inline python/ray/data/_internal/execution/streaming_executor_state.py:661]:
```suggestion
    if DataContext.get_current().use_runtime_metrics_scheduling:
        return (
            global_ok_sans_memory
            and op.metrics.average_bytes_change_per_task is not None
            and op.metrics.average_bytes_change_per_task <= 0
        )
    else:
        return global_ok_sans_memory and downstream_memory_ok
```

**Review by raulchen — APPROVED** [review]

File: python/ray/data/_internal/execution/streaming_executor_state.py (hunk 3 above)
Comment (bveeramani) [reply python/ray/data/_internal/execution/streaming_executor_state.py]: I tested this out with `test_large_e2e_backpressure`, and doing an `if` / `else` might not work. We're over the memory limit and `average_bytes_change_per_task` is initially `None`, so `consume` tasks aren't launched; as a result, we keep pulling data from the `produce` tasks until all of the `produce` tasks are complete.

Left this code as the original change for now. Hopefully that should minimize the risk, while still providing some benefit. In the long term, I think we need to devise a strategy for when metrics are `None`.
